app/services/analytics_events.rb

```diff
@@ -779,21 +779,25 @@ def idv_intro_visit
   end
 
   # @param [Boolean] success
+  # @param [Hash] proofing_components User's current proofing components
```

Comment:
Yeah, I agree, though it seems YARDoc (from what I can tell) is rather limiting in how to describe a custom shape like this. Open to suggestions if you have them, though. Maybe a `@see` tag referencing the file where the logic lives? That could also get quite repetitive to include for every method though.

Reply:
I wonder if it'd make sense to start validating the value of the proofing component fields to limit to what we're expecting, and then we could point the `@see` to that part of the code?

Reply:
Doesn't seem explicit enough to me. I still don't know when the various different values can be expected to be set

Maybe in the short term we can `@see` [this block](https://github.com/18F/identity-idp/pull/7111/files#diff-7b8e143668f47f9b24b0c28f731da268340fbc9a15a2dd287fd902044d3574daR28-R40) so at least people know which keys to expect

Reply:
Maybe we just come up with a `ProofingComponentLogging` struct and set that as the type, or just the ActiveRecord model (since it's basically a hash), and we could update our doc parsing to like go that extra step and represent that as a hash

```suggestion
  # @param [ProofingComponent] proofing_components User's current proofing components
```

Reply:
I think I like that idea. I gave it a shot in b33a721a7.

Only quibble I'd have is that the parameter isn't actually an instance of the struct, it's a hash-ified version of it.

Reply:
> Only quibble I'd have is that the parameter isn't actually an instance of the struct, it's a hash-ified version of it

Yes, but ActiveRecord objects are hash-like... so I say it still works. Like if we did pass a real `ProofingComponent` to the method we could still slice it the same way

Maybe we *should* just pass the AR model and have the class do the slicing:

```ruby
p = ProofingComponent.last
p.slice(:document_check, :document_type)
=> {"document_check"=>"mock", "document_type"=>"state_id"}
```

Reply:
I'm open to that. I'd been operating under an assumption that we want to avoid having much logic in the analytics methods themselves, though assuming the logic would at least be consistent with the documented parameters, I'm not sure it's worth worrying much over.

Reply:
Looking at this now, slicing the keys would be straightforward enough, but since there's so many of these methods accepting the parameter, it'd become quite verbose and likely require some common utility to avoid the repetition. A compromise could be to keep the implementation as-is with the Struct and then move the `.to_h` from `AnalyticsDecorator` to `AnalyticsEvents` so that the documented parameter is accurate.

Reply:
> Maybe we _should_ just pass the AR model and have the class do the slicing:

When you say "the class", can you clarify which you're referring to? I'm contemplating whether the struct should just receive the model and define the behaviors in a `to_h`, and wondering if maybe this is what you'd been suggesting all along.

Reply:
Updated in bc5b114